### PR TITLE
#5654: Removed No-Op Cores from Convs

### PR DESCRIPTION
**Pull request by sankarmanoj-tt**

### CODE REVIEW

File: tt_eager/tt_dnn/op_library/conv/multi_core_optimized_conv_sharded/optimized_conv_op_sharded_v2.cpp [hunk 1]
```diff
@@ -608,10 +608,22 @@ operation::ProgramWithCallbacks multi_core_optimized_conv_sharded_v2_(const Tens
...
         if (total_num_cores > 1) {
             std::set<CoreRange> mcast_receiver_set;
+
             if (num_cores_x > 1) {
                 mcast_receiver_set.insert(CoreRange(CoreCoord(1, 0), CoreCoord(num_cores_x - 1, 0)));
-            } if (num_cores_y > 1) {
-                mcast_receiver_set.insert(CoreRange(CoreCoord(0, 1), bottom_right_core));
+            }
+            if (num_cores_y > 1) {
+                if(total_noop_cores>0)
```
Comment (mywoodstock) [inline tt_eager/tt_dnn/op_library/conv/multi_core_optimized_conv_sharded/optimized_conv_op_sharded_v2.cpp:616]: nit: format it like:
```
if (..) {
    ...
} else {
    ...
}
```
at all places.

File: tt_eager/tt_dnn/op_library/conv/multi_core_optimized_conv_sharded/optimized_conv_op_sharded_v2.cpp [hunk 2]
```diff
@@ -1170,12 +1178,14 @@ operation::ProgramWithCallbacks multi_core_optimized_conv_sharded_v2_(const Tens
...
-                );
-                writer_ids.push_back(writer_mcast_receiver_id);
+                if(not noop_core)
+                {
+                    SetRuntimeArgs(
+                        program, writer_mcast_receiver_id, core,
+                        writer_rt_args
+                    );
+                    writer_ids.push_back(writer_mcast_receiver_id);
+                }
```
Comment (mywoodstock) [inline tt_eager/tt_dnn/op_library/conv/multi_core_optimized_conv_sharded/optimized_conv_op_sharded_v2.cpp:1188]: also update the corresponding kernels to remove the no-op checks.

**Comment by mywoodstock** [comment]
once you have all the changes, please run the post-commit CI on your branch: https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml

**Comment by sankarmanoj-tt** [comment]
I have reformatted the if...else blocks. I have removed the no_op runtime argument from the host code and the kernel. I have only modified the kernels that were used in tt_eager/tt_dnn/op_library/conv/multi_core_optimized_conv_sharded/optimized_conv_op_sharded_v2.cpp. Some kernels are unused, and I have left them like that.

**Comment by sankarmanoj-tt** [comment]
Ran post commit tests. https://github.com/tenstorrent/tt-metal/actions/runs/8874751890. Looks like all have passed.

File: tt_eager/tt_dnn/op_library/conv/multi_core_optimized_conv_sharded/optimized_conv_op_sharded_v2.cpp [hunk 3]
```diff
@@ -1043,8 +1047,6 @@ operation::ProgramWithCallbacks multi_core_optimized_conv_sharded_v2_(const Tens
                 window_outer, // window_outer
                 window_inner, // window_inner = 9 / 3, ie. read 3 width coalesced
 
-                (uint32_t) noop_core,
-
```
Comment (mywoodstock) [inline tt_eager/tt_dnn/op_library/conv/multi_core_optimized_conv_sharded/optimized_conv_op_sharded_v2.cpp:1047]: I think this arg needs to remain, else the subsequent indexing in the kernel would go wrong.

**Review by mywoodstock — CHANGES_REQUESTED** [review]

File: tt_eager/tt_dnn/op_library/conv/multi_core_optimized_conv_sharded/optimized_conv_op_sharded_v2.cpp (hunk 3 above)
Comment (sankarmanoj-tt) [reply tt_eager/tt_dnn/op_library/conv/multi_core_optimized_conv_sharded/optimized_conv_op_sharded_v2.cpp:1047]: Will revert.

**Review by mywoodstock — APPROVED** [review]

**Comment by mywoodstock** [comment]
no longer relevant.

**Pull request closed without merging**